### PR TITLE
Custom errors and BatchSizeError

### PR DESCRIPTION
**Pull request by AmbiTyga**
# What does this PR do?
The PR addresses the issue #12789.
I have added a file `custom_exceptions.py` holding a class instance to be used by `modeling_gpt2.py` to replace the assert based error with suitable exception based error.
We can add other errors as well to address the type of error happening.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
- @willfrey 
- @sgugger 
- Anyone in the community is free to review the PR once the tests have passed.

### CODE REVIEW

**Comment by willfrey** [comment]
You want to inherit from `Exception` and not `BaseException.

https://docs.python.org/3/library/exceptions.html
> The built-in exception classes can be subclassed to define new exceptions; programmers are encouraged to derive new exceptions from the Exception class or one of its subclasses, and not from BaseException. More information on defining exceptions is available in the Python Tutorial under User-defined Exceptions.

**Comment by AmbiTyga** [comment]
> Just adding a best practice note: You want to inherit from `Exception` and not `BaseException`.
> 
> https://docs.python.org/3/library/exceptions.html
> 
> > The built-in exception classes can be subclassed to define new exceptions; programmers are encouraged to derive new exceptions from the Exception class or one of its subclasses, and not from BaseException. More information on defining exceptions is available in the Python Tutorial under User-defined Exceptions.
> 
> IMO, I'd say that this could be a [ValueError](https://docs.python.org/3/library/exceptions.html#ValueError)
> 
> But that's just my opinion. If the core team has an opinion on the matter, listen to them :)

The reason behind using Custom Exception is to help users know what's the error from their side is, BatchSizeError sounds more clear and directly addresses that the problem is with the batch size.

**Comment by willfrey** [comment]
You should still inherit from `Exception` and not `BaseException`, per the official Python docs

https://docs.python.org/3/library/exceptions.html#BaseException
>The base class for all built-in exceptions. It is not meant to be directly inherited by user-defined classes (for that, use Exception). If str() is called on an instance of this class, the representation of the argument(s) to the instance are returned, or the empty string when there were no arguments.

Inheriting from BaseException can cause problems with having KeyboardInterrupt exceptions getting clobbered and having programs hang.

**Comment by AmbiTyga** [comment]
> You should still inherit from `Exception` and not `BaseException`, per the official Python docs
> 
> https://docs.python.org/3/library/exceptions.html#BaseException
> 
> > The base class for all built-in exceptions. It is not meant to be directly inherited by user-defined classes (for that, use Exception). If str() is called on an instance of this class, the representation of the argument(s) to the instance are returned, or the empty string when there were no arguments.
> 
> Inheriting from BaseException can cause problems with having KeyboardInterrupt exceptions getting clobbered and having programs hang.

Yes, I have changed that. 🤝
`BaseException` -> `Exception`

**Comment by AmbiTyga** [comment]
Hey @sgugger,
Can you give me a heads up on why is the check_code_quality failing?🙏🏽

File: src/transformers/models/gpt2/modeling_gpt2.py [hunk 1]
```diff
@@ -1117,29 +1097,21 @@ def forward(
...
             >>> encoded_choices = [tokenizer.encode(s) for s in choices]
             >>> cls_token_location = [tokens.index(tokenizer.cls_token_id) for tokens in encoded_choices]
-
             >>> input_ids = torch.tensor(encoded_choices).unsqueeze(0)  # Batch size: 1, number of choices: 2
             >>> mc_token_ids = torch.tensor([cls_token_location])  # Batch size: 1
-
             >>> outputs = model(input_ids, mc_token_ids=mc_token_ids)
             >>> lm_logits = outputs.logits
             >>> mc_logits = outputs.mc_logits
-
```
Comment (LysandreJik) [inline src/transformers/models/gpt2/modeling_gpt2.py:1142]: Please don't remove these lines, they help for readability :)

File: src/transformers/models/gpt2/modeling_gpt2.py [hunk 2]
```diff
@@ -537,11 +518,9 @@ class GPT2DoubleHeadsModelOutput(ModelOutput):
                 - gpt2-xl: 48
 
     Example::
-
             # Here is an example of a device map on a machine with 4 GPUs using gpt2-xl, which has a total of 48 attention modules:
             model = GPT2LMHeadModel.from_pretrained('gpt2-xl')
             device_map = {0: [0, 1, 2, 3, 4, 5, 6, 7, 8],
```
Comment (LysandreJik) [inline src/transformers/models/gpt2/modeling_gpt2.py:523]: Please don't remove these lines, they're necessary for doc generation and help with readability

**Review by LysandreJik — COMMENTED** [review]
Hello, thank you for your PR! I would favor using a `ValueError` here instead of a `BatchSizeError`: that's what PyTorch uses when there are errors related to dimensions, I would choose to follow their lead and raise the same kind of errors rather than a custom one.

**Comment by AmbiTyga** [comment]
Ok, I am using `ValueError` and made all other changes as well. Please go through it and let me know if there's anything else that is needed.

File: utils/style_doc.py [hunk 3]
```diff
@@ -131,15 +131,13 @@ def is_special_block(self, line):
     def init_in_block(self, text):
         """
         Returns the initial value for `self.in_block`.
-
```
Comment (LysandreJik) [inline utils/style_doc.py:134]: Could you also add back the lines here?

File: src/transformers/models/gpt2/modeling_gpt2.py [hunk 4]
```diff
@@ -822,7 +823,11 @@ def custom_forward(*inputs):
             all_hidden_states = all_hidden_states + (hidden_states,)
 
         if not return_dict:
-            return tuple(v for v in [hidden_states, presents, all_hidden_states, all_self_attentions] if v is not None)
+            return tuple(
+                v
+                for v in [hidden_states, presents, all_hidden_states, all_self_attentions, all_cross_attentions]
+                if v is not None
+            )
```
Comment (LysandreJik) [inline src/transformers/models/gpt2/modeling_gpt2.py:830]: Oh that's a good catch

**Review by LysandreJik — APPROVED** [review]
LGTM, thanks @AmbiTyga!

File: src/transformers/models/gpt2/modeling_gpt2.py (hunk 4 above)
Comment (LysandreJik) [reply src/transformers/models/gpt2/modeling_gpt2.py]: Regarding this, feel free to open a PR with the fix; that's definitely a bug that you've found!

**Comment by AmbiTyga** [comment]
Thanks, @LysandreJik. :)

**Pull request merged**